### ISSUE TITLE
Program fails to link when including utf8.h in multiple files

### ISSUE THREAD

**Issue by klusark** [nemtrif/utfcpp#31]
Using utf8cpp 3.0.2 on Archlinux

```
$ g++ --version
g++ (GCC) 8.2.1 20181127
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

file1.cc:
`#include <utf.h>`

file2.cc:
`#include <utf.h>`

```
$ g++ test1.cc test2.cc -std=c++11
/usr/bin/ld: /tmp/ccmqjhz6.o: in function `utf8::append(char32_t, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)':
test2.cc:(.text+0x0): multiple definition of `utf8::append(char32_t, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)'; /tmp/ccKIEFGd.o:test1.cc:(.text+0x0): first defined here
/usr/bin/ld: /tmp/ccmqjhz6.o: in function `utf8::utf16to8(std::__cxx11::basic_string<char16_t, std::char_traits<char16_t>, std::allocator<char16_t> > const&)':
test2.cc:(.text+0x2e): multiple definition of `utf8::utf16to8(std::__cxx11::basic_string<char16_t, std::char_traits<char16_t>, std::allocator<char16_t> > const&)'; /tmp/ccKIEFGd.o:test1.cc:(.text+0x2e): first defined here
/usr/bin/ld: /tmp/ccmqjhz6.o: in function `utf8::utf8to16(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
test2.cc:(.text+0xd2): multiple definition of `utf8::utf8to16(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'; /tmp/ccKIEFGd.o:test1.cc:(.text+0xd2): first defined here
/usr/bin/ld: /tmp/ccmqjhz6.o: in function `utf8::utf32to8(std::__cxx11::basic_string<char32_t, std::char_traits<char32_t>, std::allocator<char32_t> > const&)':
test2.cc:(.text+0x176): multiple definition of `utf8::utf32to8(std::__cxx11::basic_string<char32_t, std::char_traits<char32_t>, std::allocator<char32_t> > const&)'; /tmp/ccKIEFGd.o:test1.cc:(.text+0x176): first defined here
/usr/bin/ld: /tmp/ccmqjhz6.o: in function `utf8::utf8to32(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
test2.cc:(.text+0x21a): multiple definition of `utf8::utf8to32(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'; /tmp/ccKIEFGd.o:test1.cc:(.text+0x21a): first defined here
/usr/bin/ld: /tmp/ccmqjhz6.o: in function `utf8::find_invalid(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
test2.cc:(.text+0x2be): multiple definition of `utf8::find_invalid(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'; /tmp/ccKIEFGd.o:test1.cc:(.text+0x2be): first defined here
/usr/bin/ld: /tmp/ccmqjhz6.o: in function `utf8::is_valid(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
test2.cc:(.text+0x372): multiple definition of `utf8::is_valid(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'; /tmp/ccKIEFGd.o:test1.cc:(.text+0x372): first defined here
/usr/bin/ld: /tmp/ccmqjhz6.o: in function `utf8::replace_invalid(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char32_t)':
test2.cc:(.text+0x3ac): multiple definition of `utf8::replace_invalid(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char32_t)'; /tmp/ccKIEFGd.o:test1.cc:(.text+0x3ac): first defined here
/usr/bin/ld: /tmp/ccmqjhz6.o: in function `utf8::replace_invalid(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
test2.cc:(.text+0x458): multiple definition of `utf8::replace_invalid(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'; /tmp/ccKIEFGd.o:test1.cc:(.text+0x458): first defined here
/usr/bin/ld: /tmp/ccmqjhz6.o: in function `utf8::starts_with_bom(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
test2.cc:(.text+0x4fc): multiple definition of `utf8::starts_with_bom(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'; /tmp/ccKIEFGd.o:test1.cc:(.text+0x4fc): first defined here
/usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/8.2.1/../../../../lib/Scrt1.o: in function `_start':
(.text+0x24): undefined reference to `main'
collect2: error: ld returned 1 exit status
```

```
$ g++ test1.cc test2.cc -std=c++03
/usr/bin/ld: /usr/lib/gcc/x86_64-pc-linux-gnu/8.2.1/../../../../lib/Scrt1.o: in function `_start':
(.text+0x24): undefined reference to `main'
collect2: error: ld returned 1 exit status
```
As expected since we didn't include a main

I get a very similar error using clang as well.

Is there a way to work around this?

**Comment by nemtrif** [comment]
Ah, my bad. I added non-template free functions and forgot to declare them inline. 

Anyway, before I provide the fix, you can use the following workarounds:

1) Include utf8/checked.h instead of utf8.h and use the template functions directly.
2) Add inline to declarations of functions in cpp11.h

**Comment by nemtrif** [comment]
Fixed in v3.0.3

**Comment by klusark** [comment]
I can confirm that v3.0.3 fixed my issue. Thanks for the quick fix!
